### ISSUE TITLE
PlayerBrewedPotionEvent resulting in empty itemstack when shift-collecting

### ISSUE THREAD

**Issue by Caltinor** [MinecraftForge/MinecraftForge#8487]
**Minecraft Version:** 1.18.1

**Forge Version:** 39.0.88
**Logs:** example from log 
```
{Count:1b,id:"minecraft:potion",tag:{Potion:"minecraft:awkward"}}
{Count:0b,id:"minecraft:air",tag:{Potion:"minecraft:awkward"}}
```
first line is the output from the below reproduction steps when picking the item up with the mouse.  The second line is the output when shift-clicking to collect the potion.  note `ItemStack#serializeNBT` was used to get these outputs for maximum testing information.

**Steps to Reproduce:**
 1. print the `event.getStack()` of `PlayerBrewedPotionEvent` in a custom listener
 2. brew any potion in game
 3. collect the potion with the mouse and place in inventory to get an output similar to the first line from above log snippet
 4. repeat step 2 except collect the potion with a shift click to get an output similar to the second line from the above log snippet

**Description of issue:**
When a player collects a potion from a brewing stand by grabbing the item with the mouse and placing the item in an inventory slot, the `PlayerBrewedPotionEvent` properly returns, via `getStack()` the potion brewed.  However, if the player shift-clicks the brewed potion into their inventory, an empty itemstack of `minecraft:air` with the potion NBT is returned.

**Comment by ghost** [comment]
Forge 39.1.2
I believe this is true with the `Anvil` and `onAnvil(AnvilRepairEvent event)`

If you click the result item from the anvil the correct itemstack and nbt are returned
However, if you shift-click the result item from the anvil it returns `minecraft:air` with the correct nbt

as seen in my small project here - https://github.com/CrimsonFluff/Unbreakables  dev branch

**Comment by sciwhiz12** [comment]
Confirmed on 1.18.2-40.0.32 for both `PlayerBrewedPotionEvent` and `AnvilRepairEvent`. From investigating, it seems that this is due to a difference between the regular slot pickup code path and the "quick move" code path (shift-clicking).

The code path for the quick move shrinks the item to zero, which marks the item stack as empty but still retains its NBT data.[^1] This means that while the regular slot pickup will pass the item stack to `onTake` as expected, the quick move pickup will pass an empty item stack with the NBT data still attached.

This issue will occur for any slot which allows using quick move pickup and tries to listen to item pickups using `Slot#onTake`.

Interestingly, it seems like Mojang may be aware of this limitation of `onTake`: while it may seem on first glance that the result slots of `CartographyTableMenu` and `StonecutterMenu` also fall to this issue -- the slots' `onTake` calls `stack.getItem().onCraftedBy`, which will be the air item for a quick move pickup -- their implementations of `quickMoveStack` account for this by calling `onCraftedBy` on the item before performing the quick move via `moveItemStackTo` (which does the shrinkage described above).

[^1]: Though it retains its NBT data, the tag is of no use because the empty item stack will be overwritten by a proper one later and the item stack will not pass any `isEmpty()` checks placed before any potential `getTag()` calls.

**Comment by Shadows-of-Fire** [comment]
For brewing stands, this can be rectified by setting the stack size to 1 before the event call is passed.  The item nbt will be resurrected (as well as the item type).  This works because onTake will only have been called if there _was_ a stack in there, and so a stacksize of zero implies that it was quick-moved away beforehand.
It is also legal, because the max stack size of that slot is 1 anyway.

For anvils, the solution may be a bit more complicated, since the stacksize is not known nor guaranteed to be a specific value.

However, this can be considered to be a vanilla bug, since it means when shift-clicking, the potion passed to `CriteriaTriggers.BREWED_POTION` will always be empty.

**Comment by sciwhiz12** [comment]
> However, this can be considered to be a vanilla bug, since it means when shift-clicking, the potion passed to `CriteriaTriggers.BREWED_POTION` will always be empty.

This is incorrect: the `Potion` passed to the trigger is retrieved using `PotionUtils#getPotion`, which looks at the item stack's tag through `ItemStack#getTag()`, which returns the `CompoundTag` directly without checking whether the item stack is empty or not.[^1]

This means that the actual `Potion` in the item stack is correctly retrieved and passed to the trigger, regardless of the 'emptiness' of the item stack -- it only matters that the tag of the item stack has not yet been changed, which neither normal moving or quick moving changes.

[^1]: Interestingly enough, the only time where a method related to the item stack's tag checks the emptiness of the item stack is in `ItemStack#hasTag()` -- all other methods which check and manipulate the item stack's tag check only the nullability of the `tag` field.
